`DateTimePicker`: set PM hours correctly

ciampo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #29720

After some tweaking, the issue seems related to a bug in the [`update` function](https://github.com/WordPress/gutenberg/blob/9df736c678b2e8ef00c355e6cc9fb8092fe25484/packages/components/src/date-time/time.js#L119-L123). This function doesn't seem to take the period (`AM`/`PM`) into account, and therefore it sets the hours for the new datetime always as a number between 1 and 12 (while the new datetime expects a 24h format).

This PR uses some existing logic for the "12h to 24h format" conversion also in the `update` function

A couple of notes:
- there seems to be a visual glitch where the hours/minutes inputs flash and disappear when clicked — I will investigate this in a separate PR
- 12am is meant to represent midnight, while 12pm represents midday

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Without this PR's code applied, setting an time with the `PM` period results in the `DateTimePicker` forcing it back to the `AM` period. With this PR, setting a time in the `PM` period should work as expected.

Can be verified in:

- Storybook
- Post editor (e.g. setting a publish date in the sidebar)
- New unit test added in this PR

## Screenshots <!-- if applicable -->

Before this PR: (_when the component is using the 12 hour format and the period is set to PM, changing the hours always causes the period to automatically switch to AM_)

https://user-images.githubusercontent.com/1083581/143566626-927077a0-3ab8-4a83-9aa4-24968c0cc92d.mp4


With the fix from this PR:

https://user-images.githubusercontent.com/1083581/143477150-7c1efceb-cf23-4e7a-8396-b187fb6f40e9.mp4


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ciampo committed:
DateTimePicker: Extract `from12hTo24h` function

ciampo committed:
DateTimePicker: correctly convert hours from 12 to 24 format when updating the time

ciampo committed:
CHANGELOG

andrewserong commented on packages/components/src/date-time/time.js:
```diff
@@ -117,8 +121,16 @@ export function TimePicker( { is12Hour, currentTime, onChange } ) {
 	}
 
 	function update( name, value ) {
```
This function is passed as the `onUpdate` callback for the `dayFormat` onBlur behaviour, so unfortunately the change here introduces a regression when switching dates. To replicate, highlight the date, enter a new date and tab away from the field — for me, it's removing 12 from the date value (and causes one of the unit tests to fail).

![Kapture 2021-11-26 at 10 46 45](https://user-images.githubusercontent.com/14988353/143508506-c767bded-4e96-4779-91dc-701770b884e0.gif)

There is a similar issue with the Year field, too, as that also uses `update`, though we probably wouldn't notice it with the Month field, since there are no more than 12 months in a year 😀

Should we add another function `updateHours` that includes the logic added to `update` and then within `updateHours` call `update`? We would then pass `updateHours` to the `UpdateOnBlurAsIntegerField` around line `249`.

ciampo replied:
I should have fixed the issue by following @ntsekouras 's [inspiration](https://github.com/WordPress/gutenberg/pull/36878#discussion_r757417821) — the `from12hTo24h` function is not only called when the 12 hours format is enabled, and the input's name is `hours`.

andrewserong reviewed (commented):
Thanks for picking up this issue @ciampo! Unfortunately this introduces a regression, but it looks like it shouldn't be too hard to fix if we move some of the logic in `update` to a separate `updateHours` function. I've added some notes.

It might be nice to add an extra unit test to cover the AM/PM issue, too. We've already got a `should switch to AM correctly` test, but it looks like it only covers clicking the `AM` button, rather than the issue in #29720.

Thanks again, and happy to do more testing next week!

ntsekouras commented:
I can't reproduce the problem 😄 . Can you share a bit more testing instructions?

ciampo commented:
> I can't reproduce the problem 😄 . Can you share a bit more testing instructions?

Sure! Basically, when the component is using the 12 hour format and the period is set to `PM`, changing the hours always causes the period to automatically switch to `AM` 

https://user-images.githubusercontent.com/1083581/143566626-927077a0-3ab8-4a83-9aa4-24968c0cc92d.mp4

#29720 should also include some reproducible tests case

ntsekouras commented:
Thanks! I was changing the hour and selecting `am/pm` again, but I can see this happens when you click outside.

ciampo commented:
> Thanks! I was changing the hour and selecting `am/pm` again, but I can see this happens when you click outside.

Yes, I believe that's because the inputs in this component  "commit" their new values on blur

ntsekouras commented on packages/components/src/date-time/time.js:
```diff
@@ -117,8 +121,16 @@ export function TimePicker( { is12Hour, currentTime, onChange } ) {
+
 		// Clone the date and call the specific setter function according to `name`.
-		const newDate = date.clone()[ name ]( value );
+		const newDate = date.clone()[ name ]( adjustedValue );
```
Didn't tested much, but could this work by just adding this check:
```
if ( name === 'hours' && is12Hour && am === 'PM' ) {
	value = ( ( value % 12 ) + 12 ) % 24;
}
```
here?

ciampo replied:
Thank you for the advice, I took it as inspiration and actually applied it a few lines above to avoid calling `from12hTo24h` when not necessary!

ciampo committed:
Only apply the value adjustment when the hours input updates

ciampo committed:
Add unit test to check setting hours with PM period selected

ciampo committed:
Update unit test on the assumption that 12am should be intended as midnight

ciampo commented:
> Unfortunately this introduces a regression

Thank you for finding the time to review this and discover the regression (which was confirmed by some failing e2e/unit tests too 😅 ). That should be addressed now (more details under your inline comment).

> It might be nice to add an extra unit test to cover the AM/PM issue, too. We've already got a `should switch to AM correctly` test, but it looks like it only covers clicking the `AM` button, rather than the issue in #29720.

Definitely — I've added a new unit test to the suite.

I also think that this fix highlighted a bug in another unit test, which I corrected in this PR too — I'll add more details in an inline comment below

ciampo commented on packages/components/src/date-time/test/time.js:
```diff
@@ -47,13 +47,13 @@ describe( 'TimePicker', () => {
 
-		expect( onChangeSpy ).toHaveBeenCalledWith( '2018-12-22T12:00:00' );
+		expect( onChangeSpy ).toHaveBeenCalledWith( '2018-12-22T00:00:00' );
 		onChangeSpy.mockClear();
 
 		fireEvent.change( minutesInput, { target: { value: '35' } } );
 		fireEvent.blur( minutesInput );
 
-		expect( onChangeSpy ).toHaveBeenCalledWith( '2018-12-22T12:35:00' );
+		expect( onChangeSpy ).toHaveBeenCalledWith( '2018-12-22T00:35:00' );
```
The fix introduced in this PR caused this unit test to fail. After looking more into it, I believe that the unit test was previously wrong:

- the time picker in the test is using the 12 hours period
- the initial time set to the picker is `1986-10-18T11:00:00`, and therefore the picker automatically picks `AM` as the period
- the period is not changed explicitly throughout the test
- the `hours` input's value is changed to `12` (remember, the period is still set to `am`)
- the test currently expected `12am` to be converted to midday
- the test was passing until now because of the lack of value adjustment prior to this PR

I believe that it is correct to update the unit test in this case, since `12am` represents midnight and therefore should be converted to `00` in the 24h format

ntsekouras replied:
I think so too.

andrewserong replied:
Yes, +1 from me!

ntsekouras reviewed (approved):
Thanks @ciampo ! LGTM 💯

andrewserong reviewed (commented):
This is testing well for me, too. Thanks for fixing it up and adding in the extra test @ciampo!

ciampo commented:
I've added the backport to WP beta/RC label in case it's not too late (cc @noisysocks )

noisysocks commented:
I backported this on 29 November. You can tell if a PR was backported by checking if there's a commit in `wp/trunk` that mentions this PR. It shows up above in GitHub.

ciampo commented:
> I backported this on 29 November. You can tell if a PR was backported by checking if there's a commit in `wp/trunk` that mentions this PR. It shows up above in GitHub.

That's great! And thank you for the clarification 🙇
